### PR TITLE
S174 : new datastore lib

### PR DESCRIPTION
**Pull request by eschultink**
### Features
  - use new datastore lib, instead of GAE version

### Change implications

 - breaking change to API? **no**
 - changes dependencies?  **yes**

### CODE REVIEW

File: java/src/main/java/com/google/appengine/tools/mapreduce/MapJob.java [hunk 1]
```diff
@@ -104,7 +106,8 @@ public Value<MapReduceResult<R>> run() {
     WorkerController<I, O, R, MapOnlyMapperContext<O>> workerController = new WorkerController<>(
         jobId, new CountersImpl(), output, resultAndStatus.getHandle());
     ShardedJob<?> shardedJob =
-        new ShardedJob<>(jobId, mapTasks.build(), workerController, shardedJobSettings);
+        new ShardedJob<>(DatastoreOptions.getDefaultInstance().toBuilder()
+            .setNamespace(settings.getNamespace()).build(), jobId, mapTasks.build(), workerController, shardedJobSettings);
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/MapJob.java]: :+1: 

This is going to be a huge improvement I think

File: java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceJob.java [hunk 2]
```diff
@@ -72,6 +76,18 @@ public class MapReduceJob<I, K, V, O, R> extends Job0<MapReduceResult<R>> {
...
   @NonNull private final MapReduceSettings settings;
 
+  @Setter(onMethod = @__(@VisibleForTesting))
+  private transient Datastore datastore;
+
+  protected Datastore getDatastore() {
+    if (datastore == null) {
+      datastore = DatastoreOptions.getDefaultInstance().toBuilder()
+        .setNamespace(settings.getNamespace())
+        .build().getService();
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceJob.java:92]: Extract this to a method/builder class? Just for doing `datastore = DatastoreOptions.fromSettings(settings);`

and 
```java
static Datastore fromSettings(MapReduceSettings settings) {
return DatastoreOptions.getDefaultInstance().toBuilder()
        .setNamespace(settings.getNamespace())
        .build().getService();
}
```

As this is done several time in this class.

File: java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceServlet.java [hunk 3]
```diff
@@ -51,24 +57,48 @@
...
 
+
+  Datastore datastore;
+  MapReduceServletImpl mapReduceServletImpl = new MapReduceServletImpl(datastore);
+
+  @SneakyThrows
+  public void init() {
+    super.init();
+    if (datastore == null) {
+      datastore = DatastoreOptions.getDefaultInstance().getService();
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceServlet.java]: no `namespace` ?

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/handlers/StatusHandler.java [hunk 4]
```diff
@@ -81,15 +83,18 @@ static void handleCommand(
     response.setContentType("application/json");
     boolean isPost = "POST".equals(request.getMethod());
     JSONObject retValue = null;
+
+    //TODO: parse from request
+    Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/handlers/StatusHandler.java]: Shouldn't be associated to its namespace?

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java [hunk 5]
```diff
@@ -90,17 +74,15 @@ public class ShardedJobRunner<T extends IncrementalTask> implements ShardedJobHa
...
     return RetryerBuilder.newBuilder()
       .withWaitStrategy(WaitStrategies.exponentialWait(30_000, TimeUnit.MILLISECONDS))
       .retryIfExceptionOfType(ApiProxyException.class)
       .retryIfExceptionOfType(ConcurrentModificationException.class)
-      .retryIfExceptionOfType(DatastoreFailureException.class)
-      .retryIfExceptionOfType(CommittedButStillApplyingException.class)
-      .retryIfExceptionOfType(DatastoreTimeoutException.class)
+      //.retryIfExceptionOfType(DatastoreFailureException.class)
+      //.retryIfExceptionOfType(CommittedButStillApplyingException.class)
+     // .retryIfExceptionOfType(DatastoreTimeoutException.class)
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java:91]: Now internally handled by the new client, right? Then, I'd add this is as comment to explain why they are commented

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java [hunk 6]
```diff
@@ -196,7 +169,10 @@ private void scheduleControllerTask(Transaction tx, String jobId, String taskId,
         .param(JOB_ID_PARAM, jobId)
         .param(TASK_ID_PARAM, taskId);
     taskOptions.header("Host", settings.getTaskQueueTarget());
-    QueueFactory.getQueue(settings.getQueueName()).add(tx, taskOptions);
+
+    //Q: how can we transactionally add to queue with new library??
+    //QueueFactory.getQueue(settings.getQueueName()).add(tx, taskOptions);
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java:180]: that's the override, why is not working?

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java [hunk 7]
```diff
@@ -248,14 +226,19 @@ public void completeShard(final String jobId, final String taskId) {
       if (jobState.getStatus().getStatusCode() == DONE) {
         log.info("Calling completed for " + jobId);
         // TODO(user): consider trying failed if completed failed after N attempts
-        callCompleted(jobState);
+
+        //q: should this be same txn as above??
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java:236]: I think so; I understand that all operations done here are base on a transaction, so if something is wrong in the completion/commits, that transaction should fail too...

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/util/EntityUtils.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * class to wrap Entity with accessors that are safe for missing properties
+ *
+ * TODO: unify w com.google.appengine.tools.pipeline.impl.util.EntityUtils, when merge the two projects
+ */
+public class EntityUtils {
+  final static String IS_SHARDED_PROPERTY = "isSharded";
+
+  public static Key getKey(Entity entity, String propertyName) {
```
Comment (aperez-worklytics) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/util/EntityUtils.java:21]: Probably better with `Optional`, but that could impact in more places, so maybe for a 2nd iteration

**Review by aperez-worklytics — APPROVED** [review]
Approved; but some comments about namespace usage

File: java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceJob.java (hunk 2 above)
Comment (jlorper) [reply java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceJob.java:92]: not really follow why this getter creates an instance, doesn't that undermine the purpose of the `@Inject Datastore` ?

File: java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceJob.java [hunk 9]
```diff
@@ -204,6 +241,22 @@ private static class SortJob extends Job1<
...
+    @Setter(onMethod = @__(@VisibleForTesting))
+    private transient Datastore datastore;
+
+    @Setter(onMethod = @__(@VisibleForTesting))
+    @Inject
+    transient ShardedJobService shardedJobService;
+
+    protected Datastore getDatastore() {
+      if (datastore == null) {
+        datastore = DatastoreOptions.getDefaultInstance().toBuilder()
```
Comment (jlorper) [inline java/src/main/java/com/google/appengine/tools/mapreduce/MapReduceJob.java:253]: why can't you do ` datastore = settings.getDatastoreOptions().getService();` like in the MapJob?
Same in MergeJob.
If the settings are available, use should be similar

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/handlers/StatusHandler.java [hunk 10]
```diff
@@ -194,7 +205,7 @@ static JSONObject handleGetJobDetail(String jobId) {
       Counters totalCounters = new CountersImpl();
       int i = 0;
       long[] workerCallCounts = new long[state.getTotalTaskCount()];
-      Iterator<IncrementalTaskState<IncrementalTask>> tasks = shardedJobService.lookupTasks(state);
+      Iterator<IncrementalTaskState<IncrementalTask>> tasks = shardedJobService.lookupTasks(datastore.newTransaction(), state);
```
Comment (jlorper) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/handlers/StatusHandler.java:208]: need to close this tnx at the end?

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java [hunk 11]
```diff
@@ -119,87 +107,81 @@ public static RetryerBuilder getRetryerBuilderAggressive() {
...
 
   @VisibleForTesting
-  IncrementalTaskState<T> lookupTaskState(Transaction tx, String taskId) {
-    try {
-      Entity entity = DATASTORE.get(tx, IncrementalTaskState.Serializer.makeKey(taskId));
-      return IncrementalTaskState.Serializer.fromEntity(entity);
-    } catch (EntityNotFoundException e) {
-      return null;
+  IncrementalTaskState<T> lookupTaskState(@NonNull Transaction tx, String taskId) {
+    return (IncrementalTaskState<T>) Optional.ofNullable(tx.get(IncrementalTaskState.Serializer.makeKey(tx.getDatastore(), taskId)))
```
Comment (jlorper) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java:118]: can the `EntityNotFoundException` still happen or not? If yes, how will this behave?

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java (hunk 7 above)
Comment (jlorper) [reply java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java:236]: It looks like it should be the same...

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/pipeline/DeleteShardsInfos.java [hunk 12]
```diff
@@ -2,57 +2,59 @@
...
       String taskId = ShardedJobRunner.getTaskId(jobId, i);
-      addParentKeyToList(toDelete, IncrementalTaskState.Serializer.makeKey(taskId));
-      addParentKeyToList(toDelete, ShardRetryState.Serializer.makeKey(taskId));
+      addParentKeyToList(tx, toDelete, IncrementalTaskState.Serializer.makeKey(datastore, taskId));
+      addParentKeyToList(tx, toDelete, ShardRetryState.Serializer.makeKey(datastore, taskId));
     }
     RetryExecutor.call(
       ShardedJobRunner.getRetryerBuilder().withStopStrategy(StopStrategies.neverStop()),
-      callable(() -> DatastoreServiceFactory.getDatastoreService().delete(null, toDelete)));
+      callable(() -> tx.delete(toDelete.toArray(new Key[toDelete.size()]))));
```
Comment (jlorper) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/pipeline/DeleteShardsInfos.java:56]: don't you need to commit the tnx after?

**Review by jlorper — APPROVED** [review]
A couple of places where I think the transactions aren't committed.
Hard to grasp internal details in 30 minutes but overall seems fine.

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java (hunk 11 above)
Comment (eschultink) [reply java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java:118]: it just returns `null` in such case now directly

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/pipeline/DeleteShardsInfos.java (hunk 12 above)
Comment (eschultink) [reply java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/pipeline/DeleteShardsInfos.java:56]: yeah 👍

**Comment by eschultink** [comment]
so tests don't *actually* fully pass; CI is only picking up a subset. And haven't done full integration testing - but not really useful to treat this as independent of the Pipelines upon which it depends; there's very little use-case to having pipelines w/o map/map-reduce like functionality, esp as pipelines don't include native wrappers for BQ jobs, dataflow jobs, etc.  So not really sure what you'd want to use pipelines along to execute in a distribute, async manner ...

**Pull request merged**